File: 2552. Count Increasing Quadruplets/Solution.py
class Solution:
    def countQuadruplets(self, nums: List[int]) -> int:
        n = len(nums)
        dp = [0 for i in range(n)]
        res = 0
        for j in range(n):
            prevSmall = 0
            for i in range(j):
                if nums[j] > nums[i]:
                    prevSmall += 1
                    res += dp[i]
                else:
                    dp[i] += prevSmall
        return res

File: 2552. Count Increasing Quadruplets/Solution.java
class Solution {
    public long countQuadruplets(int[] nums) {
        int n = nums.length;
        int[] dp = new int[n];
        long res = 0;
        for (int j = 0; j < n; j++) {
            int prevSmall = 0;
            for (int i = 0; i < j; i++) {
                if (nums[j] > nums[i]) {
                    prevSmall++;
                    res += dp[i];
                } else {
                    dp[i] += prevSmall;
                }
            }
        }
        return res;
    }
}